### PR TITLE
Add Binary Allocator

### PR DESCRIPTION
**Pull request by MrQuansy**
## Description
Add Binary Allocator for object pooling and reuse of large Binary objects to reduce GC overhead.

## Features

- Object pooling for large Binary objects to minimize GC pressure
- Efficient reuse system for Binary objects
- Reduced memory allocation overhead

## Details
- Implements variable-length allocation based on Slab allocation strategy
- Features automatic eviction mechanism to prevent memory bloat
- Supports automatic start/stop capability for flexible resource management

## Next Steps
Integrate Binary Allocator into IoTDB's writing execution.

### CODE REVIEW

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/utils/binaryallocator/BinaryAllocator.java [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+      evictor.startEvictor(allocatorConfig.getDurationBetweenEvictorRuns());
+    } else {
+      state.set(BinaryAllocatorState.CLOSE);
+      this.close(false);
+    }
+  }
+
+  public PooledBinary allocateBinary(int reqCapacity) {
+    if (reqCapacity < allocatorConfig.minAllocateSize
+        | reqCapacity > allocatorConfig.maxAllocateSize) {
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/utils/binaryallocator/BinaryAllocator.java:78]: || ?

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/utils/binaryallocator/BinaryAllocator.java [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+    }
+
+    this.metrics = new BinaryAllocatorMetrics(this);
+    MetricService.getInstance().addMetricSet(this.metrics);
+
+    if (allocatorConfig.enableBinaryAllocator) {
+      state.set(BinaryAllocatorState.OPEN);
+      evictor =
+          new GCEvictor(
+              "binary-allocator-gc-evictor", allocatorConfig.getDurationEvictorShutdownTimeout());
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/utils/binaryallocator/BinaryAllocator.java:68]: do no use magic string

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/utils/binaryallocator/BinaryAllocator.java [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+      arena.evict(ratio);
+    }
+  }
+
+  public boolean isOpen() {
+    return state.get() == BinaryAllocatorState.OPEN;
+  }
+
+  public void close(boolean needReopen) {
+    if (needReopen) state.set(BinaryAllocatorState.TMP_CLOSE);
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/utils/binaryallocator/BinaryAllocator.java:132]: use {}

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/utils/binaryallocator/BinaryAllocatorState.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.commons.utils.binaryallocator;
+
+public enum BinaryAllocatorState {
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/utils/binaryallocator/BinaryAllocatorState.java:22]: image
Describe the transition relationship between these states, and what conditions trigger the transition

File: iotdb-core/datanode/src/test/java/org/apache/iotdb/db/utils/BinaryAllocatorTest.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+public class BinaryAllocatorTest {
+  @Test
+  public void testAllocateBinary() {
+    AllocatorConfig config = new AllocatorConfig();
+    config.arenaNum = 1;
+    BinaryAllocator binaryAllocator = new BinaryAllocator(config);
+    PooledBinary binary = binaryAllocator.allocateBinary(256);
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/datanode/src/test/java/org/apache/iotdb/db/utils/BinaryAllocatorTest.java:44]: use Odd number like 255 is better to test jvm allocation？

**Review by OneSizeFitsQuorum — COMMENTED** [review]
Awesome work! This is a big step forward in our performance.

But please pay attention to the code style, it seems that the overall code is written very casually and loosely at present...

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/utils/binaryallocator/BinaryAllocator.java [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+          restart();
+        }
+        return;
+      }
+
+      if (JvmGcMonitorMetrics.getInstance().getGcData().getGcTimePercentage() > 30) {
+        for (Arena arena : heapArenas) {
+          arena.evict(1.0);
+        }
+        close(true);
```
[the remark this answers is not in the document]
Comment (MrQuansy) [reply iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/utils/binaryallocator/BinaryAllocator.java:230]: Now, when GC time percentage is > 20%, the allocator will be restarted if its state is TMP_CLOSE.
BTW, I'm thinking about whether we can initialize the allocator with TMP_CLOSE state and start it when we observe serious GC overhead?

File: iotdb-core/datanode/src/test/java/org/apache/iotdb/db/utils/BinaryAllocatorTest.java (hunk 5 above)
Comment (MrQuansy) [reply iotdb-core/datanode/src/test/java/org/apache/iotdb/db/utils/BinaryAllocatorTest.java:44]: sry, I didn't get your point. I think 255 and 256 have no difference in this testcase..

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/utils/binaryallocator/AdaptiveWeightedAverage.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.commons.utils.binaryallocator;
+
+import static java.lang.Math.max;
+
+public class AdaptiveWeightedAverage {
```
[the remark this answers is not in the document]
Comment (MrQuansy) [reply iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/utils/binaryallocator/AdaptiveWeightedAverage.java:36]: Copied from jdk17

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/utils/binaryallocator/AdaptiveWeightedAverage.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+  private boolean isOld; // Enable to have enough historical data
+  private final int OLD_THRESHOLD = 100;
+
+  public AdaptiveWeightedAverage(int weight) {
+    this.weight = weight;
+    average = 0f;
+    sampleCount = 0;
+  }
+
+  public void sample(int newSample) {
```
[the remark this answers is not in the document]
Comment (MrQuansy) [reply iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/utils/binaryallocator/AdaptiveWeightedAverage.java:49]: The EWMA implementation in com.codahale.metrics introduces extra work to improve concurrent performance, which wastes additional memory and reduces update performance. We don't need to consider concurrency issues here.

BTW, the AdaptiveWeightedAverage is modified from JDK17 source codes. However, I made some modifications after discovering that memory allocation and deallocation occurs in cycles, such as before and after a request completes. Therefore, the active memory volume fluctuates significantly, so we calculate a moving average based on the peak active memory within each evictor cycle.

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

**Pull request merged**